### ISSUE TITLE
Infinite loop in zoomFromMercatorZAdjusted

### ISSUE THREAD

**Issue by mholt** [mapbox/mapbox-gl-js#12450]
**mapbox-gl-js version**: v2.11.1

**browser**: Firefox, but also happens on Webkit

### Steps to Trigger Behavior

I'm not entirely sure yet. I am still pinpointing exactly what I do that causes this, in a minimal way, but I know that it has nothing to do with the input to the map because without changing my map interactions at all, things work fine sometimes. I think it might happen if the map is not visible on the page until after page load?

What I do know is I'm calling `map.fitBounds()` that triggers it (see below).

I know that's not helpful, but I wanted to at least get this posted with the details I do currently have in case someone more familiar with the code might have a clue.

### Link to Demonstration

https://jsbin.com/lulodecebi/edit?html,js,output

~~Working on this; will update if I have further details. Unfortunately my screen recorder isn't working today (it records a video but nothing moves... sigh, Linux).~~

### Expected Behavior

The map should not have an infinite loop and the page should not freeze.

### Actual Behavior

Firefox shows "This page is slowing down Firefox" -- when I choose to debug, I see this call stack:

```
zoomFromMercatorZAdjusted (https://api.mapbox.com/mapbox-gl-js/v2.11.1/mapbox-gl.js#36)
_cameraForBoundsOnGlobe (https://api.mapbox.com/mapbox-gl-js/v2.11.1/mapbox-gl.js#36)
_cameraForBounds (https://api.mapbox.com/mapbox-gl-js/v2.11.1/mapbox-gl.js#36)
cameraForBounds (https://api.mapbox.com/mapbox-gl-js/v2.11.1/mapbox-gl.js#36)
fitBounds (https://api.mapbox.com/mapbox-gl-js/v2.11.1/mapbox-gl.js#36)
```

I'm calling the function `fitBounds()`. It works fine except in certain cases, and that's what I'm trying to isolate currently.

It's this code in transform.js:

```js
    zoomFromMercatorZAdjusted(z: number): number {
        const getZoom = (zoom) => {
            const d = this.getCameraToCenterDistance(this.projection, zoom);
            return this.scaleZoom(d / (z * this.tileSize));
        };

        let zoom = getZoom(this.zoom);
        let diff = Math.abs(zoom - getZoom(zoom));
        let lastdiff;
        while (lastdiff !== diff) {
            zoom = getZoom(zoom);
            lastdiff = diff;
            diff = Math.abs(zoom - getZoom(zoom));
        }

        return zoom;
    }
```

Stepping over repeatedly goes between the `while(...` and `zoom =` lines, never anything else.

The scopes pane reveals:

```
Function Body
​
<this>: {…}
diff: NaN
getZoom:t(t)
lastdiff: NaN
zoom: NaN
```

Inside `getZoom()`, `d` is NaN and `z` is `-2.3265482126335146e-17` (and `tileSize` is 512). Even if `d` WAS a number, this would almost be division by 0.

FWIW, these are the bounds I'm fitting:

```
_ne: Object { lng: -115.9398431, lat: 35.2138334 }
_sw: Object { lng: -115.9398431, lat: 35.2138334 }
```

(There's only 1 marker on the map in this case. Note that this usually works, so I'm still trying to isolate why it doesn't work sometimes.)

**Comment by mholt** [comment]
Sure! If it works. :) Since I haven't been able to provide a reproducer, I'd be happy to test any patches instead. I can even try using another computer with a working screen-recorder to capture the results if that's helpful.

**Comment by karimnaaji** [comment]
Thanks for the report @mholt , we should investigate that more and eventually gate it as @mourner suggested. The computations might end up lacking accuracy and not converging as expected. If you happen to hit that again giving us the parameters provided to `fitBounds` with the `lat/lon/zoom/pitch/bearing` the camera is in prior to running the method would be helpful.

**Comment by mholt** [comment]
Thanks!

Interestingly, I can print `JSON.stringify(map.getFreeCameraOptions())` to the console, but when I hit it, I get "cyclic object value". But here are the values I can get from the console when I print it non-stringified, just before calling fitBounds():

```
position:
x: 0.24375
y: 0.38660842783177896
z: 0.21731967182922918

orientation:
0: 0
1: 0
2: 0
3: 1

elevation:
(bunch of stuff, let me know what you might need from this)

zoom: 0
pitch: 0
bearing: 0
```

Input coordinates are given above.

Is that helpful?

**Comment by mholt** [comment]
I can finally reproduce it!

https://jsbin.com/lulodecebi/edit?html,js,output

**Comment by EarlOld** [comment]
I have a problem with Infinite loop too on globe projection, when try to call fitBounds (v2.11.0).

On v2.10.0 works fine

**Comment by karimnaaji** [comment]
Thanks a lot for the repro, it was super helpful in investigating that issue. I have a fix for it which both prevents `NaN` values in that function and better converges to the true expected value, the floating point accuracy was indeed an issue in that specific case.

The fix should be available in either 2.12 (currently in beta) or the release right after that.

**Comment by mholt** [comment]
Awesome, thank you!!
